test/controllers/registrations_controller_test.rb

```diff
@@ -116,4 +116,79 @@ class RegistrationsControllerTest < ActionController::TestCase
 
     assert_redirected_to course_registrations_path(@cs301)
   end
+
+  def attempt_register_user(user, role)
+    post :create, params: {
+        course_id: @cs101.id,
+        registration: {
+            username: user.username,
+            role: role
+        },
+        new_sections: [@section.crn.to_s]
+    }
+  end
+
+  test "graders cannot create registrations" do
+    mike_reg = Registration.create(course: @cs101,
+                                   user: @mike,
+                                   role: "grader",
+                                   show_in_lists: false)
+    mike_reg.save!
+    sign_in @mike
+    attempt_register_user @john, "student"
+    assert_redirected_to root_path
+    assert_match "Must be an admin, professor or assistant.", flash[:alert]
+  end
+
+  test "professors can create registrations of any role" do
+    sign_in @fred
+    Registration.roles.each do |role, num|
+      attempt_register_user @john, role
+      assert_redirected_to course_registrations_path @cs101
+      assert_nil flash[:alert]
```

Comment:
Why does this work?  If we keep re-registering John for the CS101 over and over with different roles, why should this work after the first one?

Reply:
The registration is upgraded each time - we only disallowed _downgrades_.

Reply:
Mmm, good point.  Probably worth a comment why this should work.  Do you have a comparable test (maybe models, maybe controller, dunno) that tries the reverse order and confirms that it doesn't work?

Also, it's probably worth commenting that `Registrations.roles` provides its values in order of definition, which happens to be sorted by role/permissions.  If that were to change, this test would break...

Reply:
Also, now that you're creating a new user, we're no longer re-registering John and checking that upgrade case.  I'm fine with that.